src/main/java/redis/clients/jedis/csc/AbstractCache.java

```diff
@@ -0,0 +1,223 @@
+package redis.clients.jedis.csc;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+import redis.clients.jedis.annots.Experimental;
+import redis.clients.jedis.util.SafeEncoder;
+
+/**
+ * The class to manage the client-side caching. User can provide any of implementation of this class
+ * to the client object; e.g. {@link redis.clients.jedis.csc.CaffeineClientSideCache
+ * CaffeineClientSideCache} or {@link redis.clients.jedis.csc.GuavaClientSideCache
+ * GuavaClientSideCache} or a custom implementation of their own.
+ */
+@Experimental
+public abstract class AbstractCache implements Cache {
+
+  private ClientSideCacheable cacheable = DefaultClientSideCacheable.INSTANCE; // TODO: volatile
+  private final Map<ByteBuffer, Set<CacheKey<?>>> redisKeysToCacheKeys = new ConcurrentHashMap<>();
+  private final int maximumSize;
+  private ReentrantLock lock = new ReentrantLock();
+  private volatile CacheStats stats = new CacheStats();
+
+  protected AbstractCache(int maximumSize) {
+    this.maximumSize = maximumSize;
+  }
+
+  protected AbstractCache(int maximumSize, ClientSideCacheable cacheable) {
+    this.maximumSize = maximumSize;
+    this.cacheable = cacheable;
+  }
+
+  // Cache interface methods
+
+  @Override
+  public int getMaxSize() {
+    return maximumSize;
+  }
+
+  @Override
+  public abstract int getSize();
+
+  @Override
+  public abstract Collection<CacheEntry> getCacheEntries();
+
+  @Override
+  public CacheEntry get(CacheKey cacheKey) {
+    return getFromStore(cacheKey);
+  }
+
+  @Override
+  public CacheEntry set(CacheKey cacheKey, CacheEntry entry) {
+    lock.lock();
+    try {
+      entry = putIntoStore(cacheKey, entry);
+      getEvictionPolicy().touch(cacheKey);
+      if (getEvictionPolicy().evictNext() != null) {
+        stats.evict();
+      }
+      for (Object redisKey : cacheKey.getRedisKeys()) {
+        ByteBuffer mapKey = makeKeyForRedisKeysToCacheKeys(redisKey);
+        if (redisKeysToCacheKeys.containsKey(mapKey)) {
+          redisKeysToCacheKeys.get(mapKey).add(cacheKey);
+        } else {
+          Set<CacheKey<?>> set = ConcurrentHashMap.newKeySet();
+          set.add(cacheKey);
+          redisKeysToCacheKeys.put(mapKey, set);
+        }
+      }
+      stats.load();
+      return entry;
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public Boolean delete(CacheKey cacheKey) {
+    lock.lock();
+    try {
+      boolean removed = removeFromStore(cacheKey);
+      getEvictionPolicy().reset(cacheKey);
+
+      // removing it from redisKeysToCacheKeys as well
+      // TODO: considering not doing it, what is the impact of not doing it ??
```

Comment:
you are right. my concern was regarding to runtime performance; removing them one by one ends up O(n) complexity.  
as u suggested, if that redis key does not invalidates or cachekey itself is not operated on after this removal, then it will remain in the set. It ends up in a dirty state.
thanks for pushing me to think over this.